### DISCUSSION TITLE
Reading single characters from stdin

### DISCUSSION THREAD

**Discussion by felixdoerre** [pyscript/pyscript#2152] [Proposals]
I want to write a pyscript-micropython application that runs within the in-browser terminal and processes stdin. From what I can see standard in can only be accessed with `input` which reads a whole line. However for my application I want to read and process the characters from stdin one-by-one to react to them immediately. From what I can see in the corresponding code, don't see a way to do that. The code either processes the input in a very simple "readline"-type logic, or sends it to `replProcessChar`:

https://github.com/pyscript/pyscript/blob/main/pyscript.core/src/plugins/py-terminal/mpy.js#L166

Are you aware of any workaround or could implement the possibility to read single characters from the terminal input?

**Comment by WebReflection** [comment]
the `terminal` attribute switches everything to a single char parser but also MIcroPython already has a single char parser in general and the API to deal with it is used in PyScript to create *mpy* `terminal` nodes. The *py-terminal* plugin indeed shows you exactly how we orchestrate that singel char accumulator/behavior so you should be able to implement your own py-terminal like thing and do whatever you need to do? From our side I'd rather would like to understand what's missing in our terminal than telling you to copy and paste or read our source code but it's not clear in this discussion what is your goal.

Anything else you could tell us about terminal limitation or what is it that you would like to achieve?

**Reply by felixdoerre** [reply]
So the longer story what I am trying to achieve:
Micropython (executed on a microcontroller) currently has a module called "webrepl", where the micropython REPL is exposed using a websocket, and then made available in-browser using xterm.js. The in-browser code is rather simple: https://micropython.org/webrepl/

Complementary to this, micropython comes with a utility called `mpremote` (https://github.com/micropython/micropython/tree/master/tools/mpremote), that allows accessing the REPL and helps automating tasks like file upload, or even "mounting" a local file system to the microcontroller.

I would like merge the two approaches, bringing a `webrepl`-client to `mpremote` and (using pyscript) execute `mpremote` in the browser to have the additional mpremote features available in-browser.

So in what I am imagining, I would have a pyscript-mpy running in-browser, creating a pyscript-Websocket connection to the micropython board, and then handing control over to `mpremote` (https://github.com/micropython/micropython/blob/master/tools/mpremote/mpremote/repl.py#L10), and that would then, like it currently does on PC, read single characters from stdin and send them to the micropython board-repl. 

What I am currently missing from the pyscript terminal is some API/function to do `sys.stdin.read(1)` which reads exactly one character (or byte) and does not wait for the user to press "enter" (like `input()` currently does). Sure I can patch something into pyscript but I'd rather be able to use pyscript as-is.

The other feature that might be useful is the way to make async javascript function magically appear synchronous in python. As implemented, `mpremote` expects websocket input and reading from `stdin` to be blocking. So having some way (from python) to configure `interpreter.sync` would also be useful.

**Reply by WebReflection** [reply]
I have published [micro-repl](https://github.com/WebReflection/micro-repl) which uses PyScript and it's used in various projects already so maybe you can see in there how I deal with it in PyScript? The live demo example folder is here: https://github.com/WebReflection/micro-repl/tree/main/mpy

**Reply by WebReflection** [reply]
> The other feature that might be useful is the way to make async javascript function magically appear synchronous in python. As implemented, mpremote expects websocket input and reading from stdin to be blocking. So having some way (from python) to configure interpreter.sync would also be useful.

that's possible and it's already done in mpy terminal but it requires **inevitably** a *worker* executor to do so. This part is covered in our py-terminal plugin too ... when the script is a worker it uses sync readline or sync interceptor of chars

**Reply by felixdoerre** [reply]
> I have published [micro-repl](https://github.com/WebReflection/micro-repl) which uses PyScript and it's used in various projects already so maybe you can see in there how I deal with it in PyScript?

I've now looked at that repository, but I don't think it covers what I do. From what I see, it allows to use a micropython-pyscript in-browser, or to connect to a micropython device using "webserial", where no pyscript would be involved at all. This is not what I want to do. I want to run the existing `mpremote` utility (allowing [file transfers](https://github.com/micropython/micropython/blob/42eab32a36f70840c7e670a61269fb795d43991b/tools/mpremote/mpremote/transport_serial.py#L383), mounting, handling flow-control paste mode, [mip](https://github.com/micropython/micropython/blob/42eab32a36f70840c7e670a61269fb795d43991b/tools/mpremote/mpremote/mip.py), etc), inside a mpy-pyscript environment.

>  requires inevitably a worker executor to do so.

I'm fine with using a worker, this is already the part of py-terminal, I looked at (and where the initial Link I provided, points to). In the scenario where `webrepl` is currently run, it has full control over the headers, so activating the worker mode seems easy. Using a worker is not a problem for me, the problem is that configuring the `sync` object seems not possible from within the executed python code, but only if I adjust `plugins/py-terminal/mpy.js` which I would like to avoid.

> This part is covered in our py-terminal plugin too ... when the script is a worker it uses sync readline or sync interceptor of chars

When the script uses a worker, there is a sync version of "input" (https://github.com/pyscript/pyscript/blob/main/pyscript.core/src/plugins/py-terminal/mpy.js#L144), that uses a minimalistic version of readline (https://github.com/pyscript/pyscript/blob/main/pyscript.core/src/plugins/py-terminal/mpy.js#L165), but I see not way to read individual characters (blocking) from python, and I don't see a way to configure the `sync` object, e.g. to add an additional `sync.pyterminal_read_single_char` from python.

**Reply by WebReflection** [reply]
the goal of this project is to make complicated things simple ... you are probably in the "*moar complicated*" "*rabbit hole*" and we have *polyscript* to address any underlying scenario where you can bootstrap your own thing and override the *stdio* as you wish. We don't have tons of documentation around our enabling bricks, as we don't prioritize these needs in here for the initial reason, but you can accomplish pretty much everything you want already although that'd be on the *polyscript* side of affairs ... and I hope you don't need to dig deeper in *coincident* or *sabayon* side of affairs, but the whole stack is there to make *PyScript* what it is and help developers needing more to dig into its stack.

This is to say: nothing is impossible, requirements and goals are different per each project in this stack though.

**Reply by felixdoerre** [reply]
So the result is: both, reading single characters, and customizing `sync` hooks are not possible from within `pyscript` as-is but require cooperation from the surrounding website, did I get that correctly?

Because right now I see no way to access the `interpreter` or the `sync` objects in the main page.

**Reply by WebReflection** [reply]
the way we bootstrap the interpreter is though our own io, [pyodide example here](https://github.com/pyscript/polyscript/blob/main/esm/interpreter/pyodide.js#L88) and that's way inside the stack: it's [polyscript](https://pyscript.github.io/polyscript/), not *PyScript*.

To interfere with that in *PyScript* you need to understand what our plugins do but if you want full control of the *io* in the interpreter, we don't have a story in here to enable that, but even if we did, it would be brought in here via *polyscript*.

Behind *polyscript* there's *coincident* and behind *coincident* there is *sabayon* ... and behind *sabayon* or *coincident* there is *js-proxy* ... this is to say, up to the *polyscript* project we deal with workers/main primitives, in *polyscript* we provide a library to fully take over anything you like about your interpreter.

In *PyScript* we enable with ease *almost* everything *polyscript* can provide with great control over what we need, or want, to provide but we also provide our own *Python* `stdlib` that still orchestrates what *polyscript* enables already.

I am not sure I am confusing you more than it's needed but you can see your *very low level* requirements are rather satisfied in *polyscript* project, not here, and if you manage to get what you want there, of course we'll welcome your suggestions in here too after understanding the requirements, the outcome, and the implementation.

Again, this is probably not what you wanted to hear but it's the current state of affairs and many others are (hopefully and apparently) happy with such current state. Too *low-level* control is a bit against this project goal as it exposes the project to more issues, not less, to deal with, for stuff that was likely not supposed to be "*configurable*".

I might cc @ntoll here just because he's way better than me in describing situations, and I'd be happy to learn from him, eventually, here there's some thing to change in our project ... I personally don't see it right now within this discussion.
